libs/analytics/src/widget/orderLifecycleAnalytics.ts

```diff
@@ -1,109 +1,295 @@
-// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
-const getOrderPayload = (payload: BaseOrderPayload) => ({
-  walletAddress: safeGetString(payload.order, 'owner'),
-  orderId: safeGetString(payload.order, 'uid'),
-  chainId: payload.chainId.toString(),
-  sellToken: safeGetString(payload.order, 'sellToken'),
-  buyToken: safeGetString(payload.order, 'buyToken'),
-  sellAmount: safeGetString(payload.order, 'sellAmount'),
-  buyAmount: safeGetString(payload.order, 'buyAmount'),
+export type AnalyticsPayload = Record<string, unknown>
+
+type EnrichedOrderWithTokens = EnrichedOrder & {
+  inputToken?: TokenInfo
+  outputToken?: TokenInfo
+}
+
+export function extractTokenMeta(order: Partial<EnrichedOrderWithTokens> | undefined): {
+  inputToken?: TokenInfo
+  outputToken?: TokenInfo
+} {
+  if (!order) {
+    return {}
+  }
+
+  return {
+    inputToken: order.inputToken?.address ? order.inputToken : undefined,
+    outputToken: order.outputToken?.address ? order.outputToken : undefined,
+  }
+}
+
+export function buildBaseFields(payload: BaseOrderPayload): AnalyticsPayload {
+  return {
+    walletAddress: safeGetString(payload.order, 'owner'),
+    orderId: safeGetString(payload.order, 'uid'),
+    chainId: payload.chainId.toString(),
+  }
+}
+
+type Tokenish = {
+  address?: string
+  chainId?: number
+  decimals?: number
+  logoURI?: string
+  name?: string
+  symbol?: string
+}
+
+const isValidDecimals = (decimals: number | undefined): decimals is number =>
+  typeof decimals === 'number' && Number.isInteger(decimals) && decimals >= 0
+
+const toTokenWithLogo = (meta?: Tokenish): TokenWithLogo | null => {
+  if (!meta?.address || !isValidDecimals(meta.decimals)) {
+    return null
+  }
+
+  try {
+    return TokenWithLogo.fromToken(
+      {
+        chainId: meta.chainId ?? 0,
+        address: meta.address,
+        decimals: meta.decimals,
+        symbol: meta.symbol || '',
+        name: meta.name || '',
+      },
+      meta.logoURI,
+    )
+  } catch {
+    return null
+  }
+}
+
+type RawAmount = string | number | bigint | null | undefined
+
+const normalizeRawAmount = (rawAmount: RawAmount): string | undefined => {
+  if (rawAmount === null || rawAmount === undefined) {
+    return undefined
+  }
+
+  const raw = typeof rawAmount === 'bigint' ? rawAmount.toString() : String(rawAmount)
+  const trimmed = raw.trim()
+
+  return trimmed === '' ? undefined : trimmed
+}
+
+const formatTokenUnits = (meta: Tokenish | undefined, rawAmount: RawAmount): string | undefined => {
+  const normalizedRaw = normalizeRawAmount(rawAmount)
+  if (!normalizedRaw) {
+    return undefined
+  }
+
+  const currency = toTokenWithLogo(meta)
+  if (!currency) {
+    return undefined
+  }
+
+  try {
+    const amount = CurrencyAmount.fromRawAmount(currency, normalizedRaw)
+    const formatted = formatTokenAmount(amount)
+    return formatted || undefined
+  } catch {
+    return undefined
+  }
+}
+
+const buildTokenFields = (
```

Comment:
I remember the last time we discussed how to define functions, and the decision was to always use `function` instead of `const`. @alfetopito do I remember right?